src/res/mod.rs

```diff
@@ -244,124 +257,300 @@ impl Resources {
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| Fetch {
                 inner: r.borrow(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
     /// Fetches the resource with the specified type `T` mutably.
     ///
     /// Please see `fetch` for details.
-    pub fn fetch_mut<T>(&self) -> FetchMut<T>
+    pub fn fetch_mut<T, F, P>(&self, with: F) -> FetchMut<T, P>
     where
         T: Resource,
+        F: FnOnce() -> T,
     {
-        self.try_fetch_mut().expect(RESOURCE_NOT_FOUND)
+        let res_id = ResourceId::new::<T>();
+        let r = self.resources
+            .get(&res_id)
+            .unwrap_or_else(|| self.def_res::<T>(res_id, with()));
+
+        FetchMut {
+            inner: r.borrow_mut(),
+            phantom: PhantomData,
+        }
     }
 
-    /// Like `fetch_mut`, but returns an `Option` instead of panicking in the case of the resource
-    /// being accessed mutably.
-    pub fn try_fetch_mut<T>(&self) -> Option<FetchMut<T>>
+    /// Like `fetch_mut`, but returns an `Option` instead of inserting a default value
+    /// in case the resource does not exist.
+    pub fn try_fetch_mut<T, P>(&self) -> Option<FetchMut<T, P>>
     where
         T: Resource,
     {
-        self.try_fetch_internal(TypeId::of::<T>()).map(|r| {
-            FetchMut {
+        let res_id = ResourceId::new::<T>();
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| FetchMut {
                 inner: r.borrow_mut(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
-    fn try_fetch_internal(&self, id: TypeId) -> Option<&TrustCell<Box<Resource>>> {
-        self.resources.get(&ResourceId(id))
+    fn def_res<T>(&self, res_id: ResourceId, value: T) -> &TrustCell<Box<Resource>>
+    where
+        T: Resource,
+    {
+        let mut new = self.new.lock();
+        let b: &mut Box<_> = new.entry(res_id)
+            .or_insert_with(|| Box::new(TrustCell::new(Box::new(value) as Box<Resource>)));
+
+        unsafe {
+            // This is correct because the returned can only live until the mutable borrow
+            // of `Resources`. This `Box` also lives at least until another mutable borrow,
+            // because the only way to drop it is calling `maintain` (which borrows `Resources`
+            // mutably).
+            // The raw pointer of a `Box` stays stable and valid as long as the `Box` doesn't
+            // get dropped.
+
+            let t: &TrustCell<_> = b.as_ref();
+            let raw = t as *const TrustCell<_>;
+
+            &*raw
+        }
+    }
+
+    fn get_res<T>(&self, res_id: &ResourceId) -> Option<&TrustCell<Box<Resource>>>
+    where
+        T: Resource,
+    {
+        let new = self.new.lock();
+        let b: &Box<_> = match new.get(&res_id) {
```

Comment:
You can use `?` for `Option`s.

Reply:
I know, but I don't want to use the feature yet. The rustc version bump will have to wait a few months.